dbms/src/Functions/GeoUtils.h

```diff
@@ -0,0 +1,653 @@
+#pragma once
+#include <Core/Types.h>
+#include <Core/Defines.h>
+#include <Columns/IColumn.h>
+#include <Columns/ColumnVector.h>
+#include <Core/TypeListNumber.h>
+#include <Common/typeid_cast.h>
+#include <ext/range.h>
+
+/// Warning in boost::geometry during template strategy substitution.
+#if !__clang__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
+
+#include <boost/geometry.hpp>
+
+#if !__clang__
+#pragma GCC diagnostic pop
+#endif
+
+#include <boost/geometry/geometries/point_xy.hpp>
+#include <boost/geometry/geometries/polygon.hpp>
+#include <boost/geometry/geometries/multi_polygon.hpp>
+#include <boost/geometry/geometries/segment.hpp>
+#include <boost/geometry/algorithms/comparable_distance.hpp>
+#include <boost/geometry/strategies/cartesian/distance_pythagoras.hpp>
+
+#include <array>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <algorithm>
+#include <IO/WriteBufferFromString.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+
+namespace GeoUtils
+{
+
+
+template <typename Polygon>
+UInt64 getPolygonAllocatedSize(Polygon && polygon)
+{
+    UInt64 size = 0;
+
+    using RingType = typename std::decay<Polygon>::type::ring_type;
+
+    auto addSizeOfRing = [& size](const RingType & ring) { size += sizeof(ring) + ring.capacity(); };
+
+    addSizeOfRing(polygon.outer());
+
+    const auto & inners = polygon.inners();
+    size += sizeof(inners) + inners.capacity();
+    for (auto & inner : inners)
+        addSizeOfRing(inner);
+
+    return size;
+}
+
+template <typename MultiPolygon>
+UInt64 getMultiPolygonAllocatedSize(MultiPolygon && multi_polygon)
+{
+    UInt64 size = multi_polygon.capacity();
+
+    for (const auto & polygon : multi_polygon)
+        size += getPolygonAllocatedSize(polygon);
+
+    return size;
+}
+
+template <typename CoordinateType = Float32, UInt16 gridHeight = 8, UInt16 gridWidth = 8>
```

Comment:
Strange enough, that height goes before width.
You could exchange them. Or it will be more natural, if you call it rows and columns.

But I doubt that we need different parameters for width and height (especially because it is compile time).
We could just use `grid_size`.

Reply:
Style (need `snake_case` for variables).